src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlTypes/SqlVector.cs

```diff
@@ -0,0 +1,246 @@
+
+    #region Fields
+
+    private readonly byte _elementType;
+    private readonly byte _elementSize;
+    private readonly byte[] _tdsBytes;
+    private T[] _array;
+
+    #endregion
+
+    #region Constructors
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/ctor1/*' />
+    public SqlVector(int length)
+    {
+        if (length < 0)
+        {
+            throw ADP.ArgumentOutOfRange(nameof(length), SQLResource.InvalidArraySizeMessage);
+        }
+
+        (_elementType, _elementSize) = GetTypeFieldsOrThrow();
+
+        IsNull = true;
+
+        Length = length;
+        Size = TdsEnums.VECTOR_HEADER_SIZE + (_elementSize * Length);
+
+        _tdsBytes = Array.Empty<byte>();
+        _array = Array.Empty<T>();
+        Memory = new();
+    }
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/ctor2/*' />
+    public SqlVector(ReadOnlyMemory<T> memory)
+    {
+        (_elementType, _elementSize) = GetTypeFieldsOrThrow();
+
+        IsNull = false;
+
+        Length = memory.Length;
+        Size = TdsEnums.VECTOR_HEADER_SIZE + (_elementSize * Length);
+
+        _tdsBytes = MakeTdsBytes(memory);
+        _array = memory.ToArray();
+        Memory = memory;
+    }
+
+    internal SqlVector(byte[] tdsBytes)
+    {
+        (_elementType, _elementSize) = GetTypeFieldsOrThrow();
+
+        (Length, Size) = GetCountsOrThrow(tdsBytes);
+
+        IsNull = false;
+
+        _tdsBytes = tdsBytes;
+        _array = MakeArray();
+        Memory = new(_array);
+    }
+
+    #endregion
+
+    #region Methods
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/ToString/*' />
+    public override string ToString()
+    {
+        if (IsNull)
+        {
+            return SQLResource.NullString;
+        }
+        return JsonSerializer.Serialize(Memory);
+    }
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/ToArray/*' />
+    public T[] ToArray()
+    {
+        return _array;
+    }
+
+    #endregion
+
+    #region Properties
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/IsNull/*' />
+    public bool IsNull { get; init; }
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/Null/*' />
+    public static SqlVector<T>? Null => null;
+
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/Length/*' />
+    public int Length { get; init; }
+    /// <include file='../../../../doc/snippets/Microsoft.Data.SqlTypes/SqlVector.xml' path='docs/members[@name="SqlVector"]/Size/*' />
+    public int Size { get; init; }
```

Comment:
What's the rationale behind exposing Byte Size of Vector data? Is this something that customers would use anywhere?

Reply:
This is needed for using the SqlCommand.Prepare() API for inserting the values.
The existing convention is to declare size of the SqlParameter before calling command.Prepare()
This is because, driver creates stored procedure on the sql server side and needs to know the size of the parameters for that.
There are no other scenarios.